addons/portal_rating/models/mail_message.py

```diff
@@ -2,27 +2,56 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import models
+from odoo.tools import format_datetime
 
 
 class MailMessage(models.Model):
     _inherit = 'mail.message'
 
-    def _portal_message_format(self, field_list):
-        # inlude rating value in data if requested
-        if self._context.get('rating_include'):
-            field_list += ['rating_value']
-        return super(MailMessage, self)._portal_message_format(field_list)
-
-    def _message_format(self, fnames, format_reply=True, legacy=False):
-        """ Override the method to add information about a publisher comment
-        on each rating messages if requested, and compute a plaintext value of it.
-        """
-        vals_list = super(MailMessage, self)._message_format(fnames, format_reply=format_reply, legacy=legacy)
-
-        if self._context.get('rating_include'):
-            infos = ["id", "publisher_comment", "publisher_id", "publisher_datetime", "message_id"]
-            related_rating = self.env['rating.rating'].sudo().search([('message_id', 'in', self.ids)]).read(infos)
-            mid_rating_tree = dict((rating['message_id'][0], rating) for rating in related_rating)
-            for vals in vals_list:
-                vals["rating"] = mid_rating_tree.get(vals['id'], {})
+    def _portal_get_default_format_properties_names(self, options=None):
+        """ Add request for rating information """
+        properties_names = super()._portal_get_default_format_properties_names()
+        if options and options.get('rating_include'):
+            properties_names |= {
+                'rating',
+                'rating_value',
+            }
+        return properties_names
+
+    def _portal_message_format(self, properties_names):
+        vals_list = super()._portal_message_format(properties_names)
+
+        message_to_rating = {}
+        if 'rating' in properties_names:
+            related_rating = self.env['rating.rating'].sudo().search(
+                [('message_id', 'in', self.ids)]
+            ).read(
+                ["id", "publisher_comment", "publisher_id", "publisher_datetime", "message_id"]
+            )
+            message_to_rating = {
+                rating['message_id'][0]: self._portal_message_format_rating(rating)
+                for rating in related_rating
+            }
+
+        for message, values in zip(self, vals_list):
+            rating_values = message_to_rating.get(message.id, {})
+            if rating_values:
+                values["rating"] = rating_values
+
+            if 'rating' in properties_names:
+                record = self.env[message.model].browse(message.res_id)
+                if hasattr(record, 'rating_get_stats'):
+                    values['rating_stats'] = record.sudo().rating_get_stats()
```

Comment:
Can't grep right now, but seems a bit sad that we browse and call "rating_get_stats" in a loop.

But I'm guessing it would be complicated to group records by model, and maybe "rating_get_stats" does not support multi anyway?

Reply:
I just moved existing code, but indeed I plan to optimize it with a groupby model (not sure if will be done here or in a new Pr).